lua/neotest-go/init.lua

```diff
@@ -82,60 +89,189 @@ local function get_go_package_name(_)
+    return file, tonumber(linenumber)
+  end
+  return nil, nil
+end
+
+local function is_error(lines)
+  for _, line in ipairs(lines) do
+    line = string.lower(line)
+    for _, pattern in ipairs(error_pattern) do
+      if string.match(line, string.lower(pattern)) then
+        return true
+      end
+    end
+  end
+  return false
+end
+
+local function is_test_logoutput(line)
+  if line and string.match(line, '^%s%s%s%s%s%s%s%s') then
+    return true
+  end
+  return false
+end
+
+local function get_errors_from_test(test, file_name)
+  if not test.file_output[file_name] then
+    return nil
+  end
+  local errors = {}
+  for line, output in pairs(test.file_output[file_name]) do
+    if is_error(output) then
+      table.insert(errors, { line = line - 1, message = table.concat(output, '') })
+    end
+  end
+  return errors
+end
+
 ---Convert the json output from `gotest` to an intermediate format more similar to
 ---neogit.Result. Collect the progress of each test into a subtable and add a field for
 ---the final result
 ---@param lines string[]
----@param output_file string
 ---@return table, table
-local function marshal_gotest_output(lines, output_file)
+local function marshal_gotest_output(lines)
   local tests = {}
   local log = {}
+  local testfile, linenumber
   for _, line in ipairs(lines) do
     if line ~= '' then
       local ok, parsed = pcall(vim.json.decode, line, { luanil = { object = true } })
       if not ok then
-        log = vim.tbl_map(function (l)
+        log = vim.tbl_map(function(l)
           return highlight_output(l)
         end, lines)
         return tests, log
       end
-      local output = highlight_output(sanitize_output(parsed.Output))
+      local output = highlight_output(parsed.Output)
       if output then
         table.insert(log, output)
+      else
+        testfile, linenumber = nil, nil
       end
-      local action, name = parsed.Action, parsed.Test
-      if name then
+      local action, package, test = parsed.Action, parsed.Package, parsed.Test
+      if test then
         local status = test_statuses[action]
-        -- sub-tests are structured as 'TestMainTest/subtest_clause'
-        local parts = vim.split(name, '/')
-        local is_subtest = #parts > 1
-        local parent = is_subtest and parts[1] or nil
-        if not tests[name] then
-          tests[name] = {
+
+        local testname, parenttestname = normalize_testname(package, test)
+        if not tests[testname] then
+          tests[testname] = {
             output = {},
             progress = {},
-            output_file = output_file,
+            file_output = {},
           }
         end
-        table.insert(tests[name].progress, action)
+
+        -- if a new file and linenumber is present in the current line, use this info from now on
+        -- beginn collection log data with everything after the file:linenumber
+        local new_testfile, new_linenumber = get_testfileinfo(parsed.Output)
```

Comment:
The snake case seems to be applied a bit inconsistently throughout, i.e. rather than separating the words it only separates a prefix e.g.
```suggestion
 local new_test_file, new_line_number = get_test_file_info(parsed.Output)
```